File: lesson_9_20241015/my_linked_list.js
/** Связанный список */
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }

    getData() {
        return this.data;
    }

    setData(data) {
        this.data = data;
    }

    getNext() {
        return this.next;
    }

    setNext(next) {
        this.next = next;
    }
}

class MyLinkedList {
    constructor() {
        this.head = null;
      }

     // --- First element ---

    pushToHead(data) {
        let newNode = new Node(data);
        //Make next of new Node as head
        if (this.head === null) {
            this.head = newNode;
            return;
        }
        newNode.setNext(this.head);
        //Move the head to point to new Node
        this.head = newNode;
    }

    removeFirst() {
        if (this.head === null) {
            return false;
        }
        this.head = this.head.next;
        return true;
    }

    // --- Last element ------

    pushToTail(data) {
        let newNode = new Node(data);
        // Is empty, then make the new node as head
        if (this.head === null) {
            this.head = newNode;
            return;
        }
        // Else traverse till the last node
        let last = this.head;
        while (last.next !== null) {
            last = last.next;
        }
        // Change the next of last node
        last.next = newNode;
    }

    removeLast() {
        if (this.head === null) {
            console.log("List is Empty");
            return false;
        }

        let current = this.head;
        let previous = null;

        if (current.next === null) { // one element
            this.head = null;
        } else {
            while (current.next !== null) {
                previous = current;
                current = current.next;
            }

            if (previous !== null) {
                previous.next = null; // remove last node
            }
        }
        return true;
    }

    // --- Index ----

    pushToIndex(index, data) {
        let newNode = new Node(data);
        let current = this.head;
        let jump;

        if (index < 1) {
            console.log("index out of bounds");
            return false;
        } else {
            jump = 0;
            while (jump < index - 1) {
                current = current.next;
                jump++;
            }

            newNode.next = current.next;
            current.next = newNode;

            return true;
        }
    }

    remove(index) {
        let current = this.head;
        let jump;
        if (index < 1) {
            console.log("index out of bounds");
            return false;
        } else {
            jump = 0;
            while (jump < index - 1) {
                current = current.next;
                jump++;
            }
            current.next = current.next.next;
            return true;
        }
    }

    get(index) {
        let count = -1;
        if (this.head === null) {
            return -1;
        }
        if (index < 0 || index > this.size()) {
            return -1;
        }
        let node = this.head;
        while (node !== null) {
            count++;
            if (count === index) {
                return node.getData();
            }
            node = node.getNext();
        }

        return -1;
    }

    size() {
        let node;
        let size = 0;
        for (node = this.head; node !== null; node = node.getNext()) {
            size++;
        }
        return size;
    }

    print() {
        let node = this.head;
        while (node !== null) {
            console.log(node.getData() + " ");
            node = node.getNext();
        }
    }

 }

 let lists = new MyLinkedList();
 console.log(" - pushToHead - "); 
 lists.pushToHead(5);
 lists.pushToHead(3);
 lists.pushToHead(6);
 lists.print();

 console.log(" - removeFirst - "); 
 lists.removeFirst();
 lists.print();

console.log(" - pushToTail - "); 
lists.pushToTail(7);
lists.print();

console.log(" - removeLast - "); 
lists.removeLast();
lists.print();

console.log(" - pushToIndex - "); 
lists.pushToIndex(1, 10);
lists.print();

console.log(" - remove index- "); 
lists.remove(1);
lists.print();
 
console.log("get(1) = "+lists.get(1));     
console.log("Size JS = "+lists.size());   

File: lesson_9_20241015/MyLinkedList.java
package lesson_9_20241015;

public class MyLinkedList {
    protected Node head; //ccылка на первый элемент

    public Node getHead() {
        return head;
    }

    // First element
    public void pushToHead(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            return;
        }

        newNode.setNext(head);
        head = newNode;
    }

    public boolean removeFirst() {
        if (this.head == null) {
            return false;
        }
        this.head = this.head.getNext();
        return true;
    }

    // Last element
    public void pushToTail(int data) {
        Node newNode = new Node(data);

        if (head == null) {
            head = newNode;
            return;
        }

        // Find current last element
        Node last = head;
        while (last.getNext() != null) {
            last = last.getNext();
        }

        last.setNext(newNode);
    }


   public void removeLast() {
        if (head == null) {
            System.out.println("List is Empty");
            return;
        }
        Node current = head;
        Node previous = null;

        if (current.getNext() == null) { // one element
            head = null;
        } else {
            while (current.getNext() != null) {
                previous = current;
                current = current.getNext();
            }

            if (previous != null) {
                previous.setNext(null); // remove last node
            }
        }
    }

    // --- Index ---

    public void pushToIndex(int index, int data) {
        Node newNode = new Node(data);
        Node current = head;
        int jump;

        if (index < 1) {
            System.out.println("index out of bounds");
        } else {
            jump = 0;
            while (jump < index - 1) {
                current = current.getNext();
                jump++;
            }

            newNode.setNext(current.getNext());
            current.setNext(newNode);
        }
    }

    public void remove(int index) {
        Node current = head;
        int jump;
        if (index < 1) {
            System.out.println("index out of bounds");
        } else {
            jump = 0;
            while (jump < index - 1) {
                current = current.getNext();
                jump++;
            }
            current.setNext(current.getNext().getNext());
        }
    }

    public int get(int index) {
        int count = -1;
        if (head == null) {
            return -1;
        }
        if (index < 0 || index > size()) {
            return -1;
        }

        Node node = head;
        while (node != null) {
            count++;
            if (count == index) {
                return node.getData();
            }
            node = node.getNext();
        }
        return -1;
    }

    public int size() {
        Node node;
        int count = 0;
        for (node = head; node != null; node = node.getNext()) {
            count++;
        }
        return count;
    }

    public void print() {
        Node node = head;
        while (node != null) {
            System.out.print(node.getData() + " ");
            node = node.getNext();
        }
        System.out.println();
    }

}
